post/2023/rsi-bp1/index.qmd

```diff
@@ -1,10 +1,224 @@
+Since we want to calculate [NDVI](https://en.wikipedia.org/wiki/Normalized_difference_vegetation_index), all we need to do is extract the row with NDVI's formula. 
+We can use the `short_name` column for that.
+
+``` {r calculating NDVI}
+ndvi = asi[asi$short_name == "NDVI", ]
+swarzedz_sentinel2_07_09_ndvi = calculate_indices(
+    swarzedz_sentinel2_07_09,
+    ndvi,
+    output_filename = tempfile(fileext = ".tif")
+)
+swarzedz_sentinel2_09_27_ndvi = calculate_indices(
+    swarzedz_sentinel2_09_27,
+    ndvi,
+    output_filename = tempfile(fileext = ".tif")
+)
+swarzedz_sentinel2_07_09_ndvi
+```
+
+We now can plot both rasters to see the values that were calculated for both dates.
+
+``` {r plotting NDVI rasters}
+par(mfrow = c(1, 2))
+swarzedz_sentinel2_07_09_ndvi_rast = terra::rast(swarzedz_sentinel2_07_09_ndvi)
+swarzedz_sentinel2_09_27_ndvi_rast = terra::rast(swarzedz_sentinel2_09_27_ndvi)
+terra::plot(swarzedz_sentinel2_07_09_ndvi_rast, range = c(-1, 1))
+terra::plot(swarzedz_sentinel2_09_27_ndvi_rast, range = c(-1, 1))
+```
+
+To better visualize the change in NDVI over time, we can create a difference raster.
+Values above 0 indicate that NDVI was higher on October 27 than on July 9. 
+Values below 0 indicate that the NDVI values decreased on October 27, compared to July 9. 
+Values equal to 0 mean that NDVI values haven't changed over time.
+
+``` {r calculating difference raster}
+par(mfrow = c(1, 2))
+dif = swarzedz_sentinel2_09_27_ndvi_rast - swarzedz_sentinel2_07_09_ndvi_rast
+terra::plot(dif)
+hist(dif, main = "", xlab = "NDVI")
+```
+
+We can save both rasters as one, using [`stack_rasters()`](https://permian-global-research.github.io/rsi/reference/stack_rasters.html) function.
+
+``` {r stacking result rasters}
+stack = stack_rasters(
+  c(
+    swarzedz_sentinel2_07_09_ndvi,
+    swarzedz_sentinel2_09_27_ndvi
+  ),
+  tempfile(fileext = ".vrt")
+)
+stack_rast = terra::rast(stack)
+names(stack_rast) = c("NDVI 07.09", "NDVI 09.27")
+stack_rast
+```
+
+``` {r plotting stack rasters}
+terra::plot(stack_rast, range = c(-1, 1))
+```
+
+# Conclusion
+
+The goal of this article was to demonstrate the capabilities of **rsi** package on the most common examples in remote sensing tasks.
+It's current state already shows huge potential in many remote sensing applications. 
+**rsi** provides functions which simplify processes of downloading satellite imagery and calculating spectral indices.
+If you want to learn more about **rsi** package, you can visit [GitHub repository of the project](https://github.com/Permian-Global-Research/rsi).
+
+While working on this article, I found many posibble features that could be added into **rsi**.
+An useful feature would be adding methods of simpler product filtering, based on cloud coverage and other parameters. 
+Other feature could be focused on time series analysis, by calculating indices for each available item in selected time span.
```

Comment:
ooh, interesting idea. Do you mean "NDVI for each individual file in this range", or "mean NDVI for each month in this range"? I think the first one can be done right now as, basically:

```r
get_sentinel2_imagery(
    swarzedz,
    start_date = "2023-07-01",
    end_date = "2023-09-30",
    output_filename = tempfile(fileext = ".tif"),
    composite_function = NULL
) |>
  lapply(calculate_indices, ndvi, output_filename = tempfile(fileext = ".tif"))
```

Reply:
"Mean NDVI for each <time period>" is a bit trickier:

```r
library(sf)
library(rsi)
san_antonio = st_point(c(-98.491142, 29.424349))
san_antonio = st_sfc(san_antonio, crs = "EPSG:4326")
san_antonio = st_buffer(st_transform(san_antonio, "EPSG:3081"), 5000)

ndvi <- spectral_indices() |> 
  dplyr::filter(short_name == "NDVI")

mapply(
  function(aoi, start_date, end_date, output_filename) {
    get_landsat_imagery(
      aoi = aoi, 
      start_date = start_date, 
      end_date = end_date, 
      output_filename = output_filename
    )
  },
  start_date = c("2023-09-01", "2023-10-01"),
  end_date = c("2023-09-30", "2023-10-31"),
  output_filename = replicate(2, tempfile(fileext = ".tif")),
  MoreArgs = c(aoi = list(san_antonio))
) |> 
  mapply(
    FUN = calculate_indices, 
    output_filename = replicate(2, tempfile(fileext = ".tif")),
    MoreArgs = c(indices = list(ndvi))
  )
```

That first `mapply()` shouldn't need a wrapper function... but something weird happens when `get_landsat_imagery()` passes its arguments on to `get_stac_data()`, which will take some debugging to figure out.

Also, for real applications, this is what I use my [proceduralnames](https://cran.r-project.org/web/packages/proceduralnames/index.html) package for -- `paste0(make_english_names(2), ".tif")` feels a lot nicer for real applications than `replicate(2, tempfile(fileext = ".tif")`.

Reply:
Alright, this is now fixed on `main`, and the first `mapply()` can be simplified:

```r
library(sf)
library(rsi)
san_antonio = st_point(c(-98.491142, 29.424349))
san_antonio = st_sfc(san_antonio, crs = "EPSG:4326")
san_antonio = st_buffer(st_transform(san_antonio, "EPSG:3081"), 5000)

ndvi <- spectral_indices() |> 
  dplyr::filter(short_name == "NDVI")

mapply(
  get_landsat_imagery,
  start_date = c("2023-09-01", "2023-10-01"),
  end_date = c("2023-09-30", "2023-10-31"),
  output_filename = replicate(2, tempfile(fileext = ".tif")),
  MoreArgs = c(aoi = list(san_antonio))
) |> 
  mapply(
    FUN = calculate_indices, 
    output_filename = replicate(2, tempfile(fileext = ".tif")),
    MoreArgs = c(indices = list(ndvi))
  )
```

Thanks for helping me catch another bug :smile:

Reply:
I was thinking about a feature that would work as a helper function, similar to `get_landsat_imagery()`, `get_sentinel2_imagery()` etc.. It would combine `get_stac_data()` and `calculate_indices()` into one function call. It could work like this:

- user passes in date span, indices to calculate
- rsi downloads all available products separately
- rsi calculates all selected indices for each downloaded product
- rsi returns SpatRaster, in which each band represent calculated index

So let's say i want to calculate NDVI and SAVI for the month of September, and there are 4 products available for September, the function should return raster with 8 bands, 4 for NDVI and 4 for SAVI.